### PR TITLE
feat(c/sedona-geos): Implement ST_ExteriorRing

### PR DESCRIPTION
**Pull request by LakshmiSowmya04**
Introduced the st_exteriorfile function, which returns the outer boundary line of a given geometry.
This issue is a part of https://github.com/apache/sedona-db/issues/224

### CODE REVIEW

**Comment by LakshmiSowmya04** [comment]
@petern48 can you please review my PR,thank you

File: python/sedonadb/tests/functions/test_functions.py [hunk 1]
```diff
@@ -2935,3 +2929,32 @@ def test_st_NRings(eng, geom, expected):
...
+            "LINESTRING (0 0, 10 0, 10 10, 0 10, 0 0)",
+        ),
+        (
+            "MULTIPOLYGON (((0 0, 1 0, 1 1, 0 1, 0 0)), ((10 10, 20 10, 20 20, 10 20, 10 10)))",
+            None,
+        ),
+        ("GEOMETRYCOLLECTION(POINT(1 1), POLYGON((0 0, 1 0, 1 1, 0 0)))", None),
+        (
+            "POLYGON Z ((0 0 1, 1 0 1, 1 1 1, 0 1 1, 0 0 1))",
+            "LINESTRING Z (0 0 1, 1 0 1, 1 1 1, 0 1 1, 0 0 1)",
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_functions.py]: Can you also add `POLYGON M` and `POLYGON ZM` cases? I think your current code might fail these, at the moment, since I see logic for `has_z` instead of using our usual pattern.

You can `Ctrl + f` in this `test_functions.py` file and search for `POLYGON M` and `POLYGON ZM` to find examples of test cases.

File: python/sedonadb/tests/functions/test_functions.py [hunk 2]
```diff
@@ -2904,12 +2904,6 @@ def test_st_numpoints(eng, geom, expected):
         ("MULTIPOINT EMPTY", 0),
         ("LINESTRING EMPTY", 0),
         ("MULTILINESTRING EMPTY", 0),
-        ("MULTIPOINT ((0 0), (1 1))", 0),
-        ("MULTILINESTRING ((0 0, 1 1), (2 2, 3 3))", 0),
-        ("POINT EMPTY", 0),
-        ("MULTIPOINT EMPTY", 0),
-        ("LINESTRING EMPTY", 0),
-        ("MULTILINESTRING EMPTY", 0),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_functions.py:3342]: Thanks for doing this. For other readers, this is just clean up from a previous PR, as explained [here](https://github.com/apache/sedona-db/pull/387#discussion_r2581685826)

File: python/sedonadb/tests/functions/test_functions.py [hunk 3]
```diff
@@ -2935,3 +2929,32 @@ def test_st_NRings(eng, geom, expected):
         f"SELECT ST_NRings({geom_or_null(geom)})",
         expected,
     )
+
+
+@pytest.mark.parametrize("eng", [SedonaDB, PostGIS])
+@pytest.mark.parametrize(
+    ("geom", "expected"),
+    [
+        (None, None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_functions.py]:
```suggestion
        (None, None),
        ("POINT EMPTY", None),
        ("LINESTRING EMPTY", None),
        ("POLYGON EMPTY", 0),
        ("MULTIPOINT EMPTY", None),
        ("MULTILINESTRING EMPTY", None),
        ("MULTIPOLYGON EMPTY", None),
        ("GEOMETRYCOLLECTION EMPTY", None),
```

As always, I want to test empty versions of each of the geometry types

File: python/sedonadb/tests/functions/test_functions.py [hunk 4]
```diff
@@ -2935,3 +2929,32 @@ def test_st_NRings(eng, geom, expected):
...
+        ("POINT (1 2)", None),
```
Comment (petern48) [inline python/sedonadb/tests/functions/test_functions.py]:
```suggestion
        ("POINT (1 2)", None),
        ("MULTIPOINT ((0 0), (1 1))", None),
```

I prefer to also test a non-empty version of each geometry type as well. You have all except multipoint already.

File: c/sedona-geos/src/st_exteriorring.rs [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+    fn invoke_batch(
+        &self,
+        arg_types: &[SedonaType],
+        args: &[ColumnarValue],
+    ) -> Result<ColumnarValue> {
+        let executor = GeosExecutor::new(arg_types, args);
+        let mut builder = BinaryBuilder::with_capacity(
+            executor.num_iterations(),
+            100 * executor.num_iterations(),
```
Comment (petern48) [inline c/sedona-geos/src/st_exteriorring.rs]:
```suggestion
            WKB_MIN_PROBABLE_BYTES * executor.num_iterations(),
```

To figure out how to set values like this, I recommend looking at other functions that return a geometry like `st_centroid.rs`. e.g. [here](https://github.com/apache/sedona-db/blob/c48dfaa06858b5fe3ddf8d68115a890fe222cab1/c/sedona-geos/src/st_centroid.rs#L53-L56)

You'll need to add the import at the top too:
```rust
use sedona_geometry::wkb_factory::WKB_MIN_PROBABLE_BYTES;
```

File: c/sedona-geos/src/st_exteriorring.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+            if geos_geom.has_z().unwrap_or(false) {
+                writer.set_output_dimension(OutputDimension::ThreeD);
+            }
+
+            let wkb = writer.write_wkb(&ring).map_err(|e| {
+                DataFusionError::Execution(format!("Failed to convert to wkb: {e}"))
+            })?;
+
+            builder.append_value(wkb);
```
Comment (petern48) [inline c/sedona-geos/src/st_exteriorring.rs]: While it seems like this works (though I think it would fail for `M` and `ZM` cases), we should follow existing code patterns instead of introducing this less concise and somewhat more complex logic.

I encourage you to look at `st_centroid.rs` to see how you can implement the wkb writing more cleanly. Specifically look [here](https://github.com/apache/sedona-db/blob/c48dfaa06858b5fe3ddf8d68115a890fe222cab1/c/sedona-geos/src/st_centroid.rs#L78-L82) and how that's called [here](https://github.com/apache/sedona-db/blob/c48dfaa06858b5fe3ddf8d68115a890fe222cab1/c/sedona-geos/src/st_centroid.rs#L59-L63)

File: c/sedona-geos/src/st_exteriorring.rs [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        match result_z {
+            ScalarValue::Binary(Some(bytes)) | ScalarValue::LargeBinary(Some(bytes)) => {
+                let expected_bytes = geos_wkt_to_wkb(z_wkt_out);
+                assert_eq!(
+                    bytes, expected_bytes,
+                    "Z-coordinates were not preserved correctly"
+                );
+            }
+            _ => panic!("Expected Binary result for 3D test"),
+        }
```
Comment (petern48) [inline c/sedona-geos/src/st_exteriorring.rs]: (Genuine question) Do we really need this extra logic for testing `POLYGON Z` here? I would've thought we could use it in the `input_wkt` and `expected` arrays below.

Let's remove this complexity by moving this test case into the arrays below. If that doesn't work. Let me know.

**Review by petern48 — COMMENTED** [review]
Nice. While it seems to work, we need to add some other test cases and do a bit of revising to clean up some unnecessary complexity. It's close, not far off from where it needs to be for us to merge it.

**Comment by petern48** [comment]
Could you also fix the PR title and mention the issue in the PR description, according to my previous comment before: https://github.com/apache/sedona-db/pull/386#issuecomment-3591890245. You should generally try to do it whenever you first open the PR. If you don't do so, it increases the odds of someone else taking up that work too, which would eventually lead to someone (either you or them) going through the work just for their PR to not be merged.

If you're afraid of linking the issue earlier because you're not sure you can 100% commit, it's totally okay to link it anyways and later decide to close your PR if you decide you don't want to follow through with it (e.g. if it's too hard or you lose interest). Nothing ever locks you in to being *required* to finish something. But please do link the issue as soon as you open the PR. For this PR, we're already close to being done anyway.

File: python/sedonadb/tests/functions/test_functions.py (hunk 1 above)
Comment (LakshmiSowmya04) [reply python/sedonadb/tests/functions/test_functions.py]: can I ask how you figured out we have to use the polygon M and polygon ZM cases for this? I cannot see it in the documentation.

Comment (petern48) [reply python/sedonadb/tests/functions/test_functions.py]: When it comes to geospatial knowledge. You just learn overtime that there's four possible dimensions. X and Y (required), and Z and M (both optional). These cases always existed, but they were less interesting cases for your previous functions which only returned counts of rings. Whether an M or Z dimension existed or not is unlikely to fail for those "counting" functions. We could totally add them, but they were less interesting or unlikely to break things.

For this ST_ExteriorRing, it doesn't just count rings, it **returns them**. Specifically, the returned geometry should also have Z and M dimensions if their inputs did. It's certainly possible for underlying implementations to forget to implement Z and M support. In that case, the function might right `LINESTRING` except missing the Z and M dimensions. That's why Z and M tests are more interesting for this function, hence we want to test them and make sure it doesn't test the wrong input.

In contrast, if ST_NumRings completely disregarded Z and M dimensions, it probably would still return the right count, since it doesn't really care about the exact contents of the rings, just the number of them. This intuition isn't quite something contributors would be expected to reason about, so don't worry too much if you can't follow the logic.
